File: 백준/Silver/12026. BOJ 거리/BOJ 거리.py
N = int(input())
blocks = list(input())

dp = [1e9] * N
dp[0] = 0

for i in range(N):
    if dp[i] == 1e9:
        continue

    if blocks[i] == 'B':
        for j in range(i, N):
            if blocks[j] == 'O':
                dp[j] = min(dp[j], dp[i] + (j-i)**2)
    elif blocks[i] == 'O':
        for j in range(i, N):
            if blocks[j] == 'J':
                dp[j] = min(dp[j], dp[i] + (j-i)**2)
    elif blocks[i] == 'J':
        for j in range(i, N):
            if blocks[j] == 'B':
                dp[j] = min(dp[j], dp[i] + (j-i)**2)

if dp[N-1] == 1e9:
    print(-1)
else:
    print(dp[N-1])

File: 백준/Silver/12026. BOJ 거리/BOJ 거리.java
import java.io.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        String blocks = br.readLine();

        int[] dp = new int[N];
        dp[0] = 0;
        for (int i = 1; i < N; i++) {
            dp[i] = Integer.MAX_VALUE;
        }

        for (int i = 0; i < N; i++) {
            if (blocks.charAt(i) == 'B') {
                for (int j = i+1; j < N; j++) {
                    if (blocks.charAt(j) == 'O') {
                        int tmp = Integer.MAX_VALUE;
                        if (dp[i] != Integer.MAX_VALUE) {
                            tmp = dp[i] + (j - i) * (j - i);
                        }
                        dp[j] = Math.min(dp[j], tmp);
                    }
                }
            } else if (blocks.charAt(i) == 'O') {
                for (int j = i+1; j < N; j++) {
                    if (blocks.charAt(j) == 'J') {
                        int tmp = Integer.MAX_VALUE;
                        if (dp[i] != Integer.MAX_VALUE) {
                            tmp = dp[i] + (j - i) * (j - i);
                        }
                        dp[j] = Math.min(dp[j], tmp);
                    }
                }
            } else if (blocks.charAt(i) == 'J') {
                for (int j = i+1; j < N; j++) {
                    if (blocks.charAt(j) == 'B') {
                        int tmp = Integer.MAX_VALUE;
                        if (dp[i] != Integer.MAX_VALUE) {
                            tmp = dp[i] + (j - i) * (j - i);
                        }
                        dp[j] = Math.min(dp[j], tmp);
                    }
                }
            }
        }

        if (dp[N-1] == Integer.MAX_VALUE) {
            System.out.println(-1);
        } else {
            System.out.println(dp[N-1]);
        }
    }
}
